File: exercism/kotlin/nucleotide-count/src/main/kotlin/DNA.kt
private val legalNucleotides = "ACGT"

class DNA(val dna: String) {

  init {
    if (!dna.all { it in legalNucleotides }) {
      throw IllegalArgumentException("Illegal character in DNA sequence")
    }
  }

  fun count(nucleotide: Char): Int =
    if (nucleotide in legalNucleotides) {
      dna.count { it == nucleotide }
    } else {
      throw IllegalArgumentException("Illegal nucleotide character.")
    }

  val nucleotideCounts: Map<Char, Int>
    get() = legalNucleotides.map { it to count(it) }.toMap()

}

File: exercism/haskell/nucleotide-count/src/DNA.hs
{-# LANGUAGE TupleSections #-}

module DNA (count, nucleotideCounts) where

import Data.Map (Map, fromListWith, findWithDefault)
import Data.Bifunctor (second)

type Nucleotide = Char

nucleotides :: [Nucleotide]
nucleotides = "ATCG"

valid :: Nucleotide -> Bool
valid c = elem c nucleotides

count :: Nucleotide -> [Nucleotide] -> Either String Integer
count c
  | valid c = second (findWithDefault 0 c) . nucleotideCounts
  | otherwise = const $ Left "Illegal nucleotide character."

nucleotideCounts :: [Nucleotide] -> Either String (Map Nucleotide Integer)
nucleotideCounts ss
  | all valid ss = Right $ fromListWith (+) (defaults ++ map (, 1) ss)
  | otherwise = Left "Illegal character in dna sequence."
  where defaults = map (, 0) nucleotides
